Add tilt cover

mtl010957 opened the pull request:
## Description:
This is my initial working version of adding position and tilt setting capabilities to the existing template cover component. This version works in optimistic mode but I think it is not working properly with lambdas to generate position and tilt states. Please review and set me on the right path.

**Related issue (if applicable):** fixes [feature-request #42](https://github.com/esphome/feature-requests/issues/42)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#207
**Pull request in [esphome-core](https://github.com/esphome/esphome-core) with C++ framework changes (if applicable):** N/A

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphomedocs](https://github.com/OttoWinter/esphomedocs).

mtl010957 commented:
@OttoWinter What is your vision for how to handle this now? I think I understand the new approach but I have questions. I see the template platform, and the new ensdstop and time_based, but I don't see how any of them supports tilt yet. Should I work on creating a new tilt platform based on the new structure? I'll help any way I can, just need a little direction.

Thanks.

OttoWinter commented:
@mtl010957 So yes the way I think this should continue:

1. Cramming it all into the template platform will probably not be the best way to go - I don't think it's possible for any template platform syntax to be easy to use.

2. The way I envision this (and what I've done with the two new platforms) is to have many different cover platforms - one for each possible type of setup. These platforms should try to solve one task and limit themselves to that. That way we don't end up with monstrous platforms that try to solve everything.

So for new cover platforms process will be: Find lowest common denominator for capabilities and way of operation but don't over-generalize. Then try to solve that task well.

I'm planning to also add a custom cover platform so that some sort of templating capability is available.

Idk, maybe this is also a wrong approach and we'll end up with many cover types that are very similar - but I'd really like to avoid over-generic integrations and this could be a test to see how it works out (for future of esphome).

mtl010957 commented:
@OttoWinter Thanks, your strategy of many small instead of one big complex platform makes perfect sense to me. I will work on a tilt only platform, since that's the type of cover hardware I have to test with. I'll try to re-purpose this PR or if that fails I'll close this and open a new one.

mtl010957 committed:
Adding new tilt cover platform
Minor cleanups
Adding tilt cover Python code
Added setup to handle restore
Added test for tilt cover

mtl010957 commented:
Reopened with a narrower focus on just tilt and a fresh start on the new single repository architecture.

mtl010957 committed:
Added speed control feature to avoid mechanical stress
Added speed settings to test

mtl010957 commented:
@OttoWinter This is stable and ready for review now, please let me know if anything needs repair. Thanks!

mtl010957 committed:
Add units to speeds and handle same as stepper

OttoWinter commented on esphome/components/cover/cover.h:
```diff
@@ -165,6 +170,8 @@ class Cover : public Nameable {
   virtual void control(const CoverCall &call) = 0;
   virtual std::string device_class();
+  float tilt_closed_value_{0.0};
+  float tilt_opened_value_{100.0};
```
ESPHome uses fixed values for closed/open - 0.0 means closed and 1.0 means open. I don't see a reason we should have these values?

mtl010957 replied:
These are actually values that are just sent to HA to tell it what value to set when the user pushes the open or close arrows in HA. That's why I had to add them to the base cover to get them into MQTT. If these don't get sent then the HA open and close arrows are kind of useless because most tilt covers would not have to be driven all the way to the end stops to open and close them. In my case, I call "closed" = 5% (louvers roughly 10 degrees off vertical) and "open" = 45% (louvers nearly horizontal). 0% might be closed but 100% is actually closed the other way so that really won't work.

OttoWinter commented on esphome/components/mqtt/mqtt_cover.cpp:
```diff
@@ -92,10 +102,20 @@ bool MQTTCoverComponent::publish_state() {
+      state_s = "open";
+    } else if (this->cover_->tilt <= this->cover_->get_tilt_closed_value()) {
+      state_s = "closed";
+    }
```
This is wrong. Tilt is its own thing - the "main" state of a cover is only based on its position (open/closed). Tilt is its own additional property.

mtl010957 replied:
I agree, but it leaves a tilt only cover in HA looking kind of strange because it never gets a valid position so the state never changes from "undefined". In the case of a cover that has both tilt and position then I'd agree the open/closed state should be defined by the position. If the HA component had the ability to show the tilt as the state of a tilt only cover it would be better.

OttoWinter commented on esphome/components/tilt/cover.py:
```diff
@@ -0,0 +1,50 @@
+
+CONFIG_SCHEMA = cover.COVER_SCHEMA.extend({
+    cv.GenerateID(): cv.declare_id(TiltCover),
+    cv.Required(CONF_TILT_ACTION): automation.validate_automation(single=True),
```
I'm not sure what this integration is for...

Why is `tilt` used here? The `position` property perfectly describes what this component does: A value in the range 0.0 to 1.0 that can be controlled by the user.

Isn't this more of a template cover with position template support?

mtl010957 replied:
As I understand how HA handles it, tilt is separate from position. For example, you could have a set of Venetian blinds that can have the tilt of the louvers controlled as well as the up and down position of the whole assembly of louvers. In that case there must be both tilt and position properties available separately from one another. I tried the template cover but it can't be coerced into telling HA that you are actually trying to control the tilt so the HA frontend never looks right.

mtl010957 replied:
Maybe some pictures will help. There are (at least) four different ways covers can be represented in HA. There's the simple open and closed cover that has no other state:
![cover-details](https://user-images.githubusercontent.com/7416838/58349054-443eb100-7e30-11e9-8215-98835bb0b811.png)

Then if you have a cover that can be set to a specific position it looks like this:
![cover-position-details](https://user-images.githubusercontent.com/7416838/58349147-7fd97b00-7e30-11e9-9794-0df548647db3.png)
It still has the notion of "open" and "close" (poke the arrows), but it also has the position slider. Open and close might be at some position other than 100% and 0% in this case, there are two settings that HA supports to define those limits (like I did with the tilt).

If you have a cover that can only control tilt, it looks like this:
![cover-tilt-details](https://user-images.githubusercontent.com/7416838/58349231-bdd69f00-7e30-11e9-9448-b92115444cab.png)
Note the arrows are angled and the slider says tilt position.

Then if you have the real high end setup that can control both tilt and position it looks like this:
![cover-position-tilt-details](https://user-images.githubusercontent.com/7416838/58349295-e494d580-7e30-11e9-9aa3-b74e020eca7e.png)

I agree it seems more complicated than it needs to be, but that's how the implementation works as far as I can tell right now.

Hope this makes sense, thanks.

OttoWinter replied:
Yeah I understand how it looks like in HA - and I think I also understand why you want it to be a "tilt-only" cover in this case.

However, I don't want frontend things to influence ESPHome's architecture - and the thing is that the way covers were designed architecturally, they always have some UP/CLOSED state; and tilt is an *independent* value that components can optionally provide.

So regarding this tilt cover, I think a template cover with position support would perfectly model the use-case. Yes you won't get the HA icons, but I mean who cares about that :)

I don't know what hardware you're using this with, but I think the `time_based` cover would even work directly for you.

The other thing is that a template cover platform should not do something like defining opening/close speeds. Template platforms should really just forward all data *directly* to the templates. For other cases the secondary non-lambda based platforms like time_based should be used.

mtl010957 replied:
I'll give the template cover and/or time_based cover a try and see if I can make it work. It will require manual HA configuration to get the frontend look I want rather than using discovery but that's not necessarily a bad thing.

mtl010957 replied:
Maybe I'm confused but I don't see how the template or time_based platforms can take an arbitrary position value commanded by the frontend (moving the tilt position slider manually for instance) and set an output based on that. Can you show me an example?

My setup needs a float position number coming in from the frontend to be sent to an output.set_level to set a PWM duty cycle using an esp8266_pwm output driving a servo motor.

Thanks.

mtl010957 replied:
@OttoWinter : I certainly understand your concerns with creating an integration that's not necessary and would have to be supported and maintained. I couldn't figure out how the existing template cover could work, but I've created a configuration that can basically replicate what the tilt cover did without using anything new, just a brute force use of MQTT publish and subscribe and a heavy dose of lambdas:

substitutions:
  devicename: nm05_sv1
  upper_devicename: NM05_SV1

esphome:
  name: ${devicename}
  platform: ESP8266
  board: nodemcuv2
  on_boot:
    priority: -80
    then:
      - delay: 2s
      - logger.log:
          format: "Sending MQTT Discovery Message"
          tag: "esphome"
      - mqtt.publish_json:
          topic: "homeassistant/cover/${devicename}/${devicename}_cover/config"
          retain: True
          payload: |-
            root["name"] = "${upper_devicename} Cover Tilt";
            root["state_topic"] = "${devicename}/cover/${devicename}_cover/state";
            root["availability_topic"] = "${devicename}/status";
            root["unique_id"] = "ESPcover${devicename}_cover_tilt";
            root["tilt_status_topic"] = "${devicename}/cover/${devicename}_cover/tilt/state";
            root["tilt_command_topic"] = "${devicename}/cover/${devicename}_cover/tilt/command";
            root["tilt_closed_value"] = 5;
            root["tilt_opened_value"] = 45;
            JsonObject& device = root.createNestedObject("device");
            device["identifiers"] = get_mac_address();
            device["name"] = "${devicename}";
            device["sw_version"] = "esphome v" ESPHOME_VERSION " " + App.get_compilation_time();
            device["model"] = ARDUINO_BOARD;
            device["manufacturer"] = "espressif";
wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
mqtt:
  broker: !secret mqtt_broker
  username: !secret mqtt_username
  password: !secret mqtt_password
# Enable logging
logger:
  level: DEBUG
ota:
globals:
  - id: tilt
    type: int
sensor:
  - platform: mqtt_subscribe
    name: "${upper_devicename} Cover Tilt"
    id: mqtt_cover_tilt
    topic: "${devicename}/cover/${devicename}_cover/tilt/command"
    qos: 0
    discovery: False
    on_value:
      then:
        - logger.log:
            format: "Setting cover tilt commanded value"
            tag: "sensor.mqtt_subscribe.cover.tilt"
        - while:
            condition:
              lambda: |-
                return (int(x) != id(tilt));
            then:
              - lambda: |-
                  if (int(x) > id(tilt)) {
                    id(tilt) += 1;
                  } else if (int(x) < id(tilt)) {
                    id(tilt) -= 1;
                  }
              - delay: 100ms
              - output.set_level:
                  id: nodemcu_servo
                  level: !lambda |-
                    return id(tilt)/100.0f;
              - mqtt.publish:
                  topic: "${devicename}/cover/${devicename}_cover/tilt/state"
                  payload: !lambda "return to_string(id(tilt));"
                  qos: 0
                  retain: True
        - mqtt.publish:
            topic: "${devicename}/cover/${devicename}_cover/state"
            payload: !lambda |-
              if (x > 5) {
                return "open";
              } else {
                return "closed";
              }
            qos: 0
            retain: True
output:
  - platform: esp8266_pwm
    pin:
      number: D5
      inverted: False
    frequency: 50 Hz
    max_power: 0.135
    min_power: 0.035
    id: nodemcu_servo

This will work for my needs for now so this PR can be closed if it doesn't fit the strategy better than the way I solved it here. I think the speed control code might be a good thing to isolate into its own component for use in other situations, it seems like something that could be reused and it's a bit tricky to implement correctly. Please let me know what you think.


Thanks.

mtl010957 committed:
Lint

mtl010957 committed:
More lint

mtl010957 commented:
Closing this one in favor of the much simpler approach enabling tilt in the Template Cover [PR#577](https://github.com/esphome/esphome/pull/577)
